### DISCUSSION TITLE
OMPython issue with readRealMatrix

### DISCUSSION THREAD

**Discussion by thomash303** [OpenModelica/OpenModelica#13337] [General]
I am using OMPython on my windows machine. I have had no problems running the bouncing ball and several custom scripts; however, when I am trying to use a model that reads external data from a .mat file using readRealMatrix it throws an error (Error: Error building simulator.). I also get this error modelname.mo:_er_12jac.c:1682:49: error: use of undeclared identifier '_omcQ_24TMP_5F8350'. As soon as I remove the sub models which use the readRealMatrix the model will translate again. I should also note, there are no issues when running this exact code in the OpenModelia editor.

I appreciate any help,

Thanks!

**Comment by arun3688** [comment]
@thomash303 can you provide your model and also the script on how your are reading the external data from .mat file, so that it will be easier to reproduce the problem and help you

**Reply by thomash303** [reply]
@arun3688  I can't share the full model, but it is essentially just two instances of a custom force block acting on the body element. 

An example of the way I'm reading the data is as follows: 
partial model multibodyData
  extends DataImport.filePath;
protected
  parameter Real nbodies_read = scalar(Modelica.Utilities.Streams.readRealMatrix(fileName, "hydro.bodies.Nb", 1, 1)) "Number of bodies";
  parameter Integer nbodies = integer(nbodies_read) "Number of bodies";
  parameter Integer bodyDoF = 6 "Degrees-of-Freedom per body";
  parameter Integer nDoF = nbodies*bodyDoF "Total Degrees-of-Freedom";     
end multibodyData;

When looking at the build log I get this error modelName_12jac.c:1835:49: error: use of undeclared identifier '_omcQ_24TMP_5F6880'.

Version and OS
Python Version: 3.13
OMPython Version: Unsure, whatever came with OpenModelica Installation
OpenModelica Version: v1.24.0
OS: [e.g. Windows 10, 64 bit]: Windows 11, 64 bit

**Comment by adrpo** [comment]
Possible duplicate of #13315

**Comment by thomash303** [comment]
@arun3688 I apologize for writing again, but do you have an idea what might be causing this problem?

**Comment by arun3688** [comment]
@thomash303 Unfortunately i did not have time to look into this, i am busy with other works, but does the normal simulation works from OMEdit ?

**Reply by thomash303** [reply]
@arun3688 The normal simulation works fine in OMEdit, that is why I am so confused about getting this error.

**Reply by thomash303** [reply]
I also got permission and I can share the source code, I have put it on this repo (https://github.com/thomash303/OceanEngineeringToolbox) on the refinement branch. To run the simulation, the Example.multibodyWEC. The directory in the filePath just has to be the path to the .mat file to execute a simulation.

**Reply by arun3688** [reply]
@thomash303 the model in the example is the full model to be tested ?

**Reply by thomash303** [reply]
@arun3688 Yes, you should not need to change anything to run the simulation except for the directory.

**Reply by arun3688** [reply]
@thomash303 can you please tell me which filename should i use to read from the repository ? for the below line
`parameter Real nbodies_read = scalar(Modelica.Utilities.Streams.readRealMatrix(fileName, "hydro.bodies.Nb", 1, 1)) "Number of bodies";
`

**Reply by thomash303** [reply]
![image](https://github.com/user-attachments/assets/00379598-0653-4808-9fce-0156f744042e)
![image](https://github.com/user-attachments/assets/cd11203e-fb35-425b-854a-e674310f45bc)

@arun3688  On the top level of the model, change the filePath to the directory for the repo. You should not need to change the file name.

**Reply by arun3688** [reply]
@thomash303 Thanks i can also able to simulate the model in OMEdit, So it should work from OMPython also, i will test it and get back to you

**Reply by arun3688** [reply]
@thomash303  i did a quick testing in OMPython and it works fine for me, see below scripting

```
>>> from OMPython import OMCSessionZMQ
>>> omc = OMCSessionZMQ()
>>> omc.sendExpression("getVersion()")
'OpenModelica v1.25.0-dev-226-gef9c2965fc (64-bit)'
>>> omc.sendExpression("loadFile(\"C:/OPENMODELICAGIT/matlab/OceanEngineeringToolbox/OET.mo\")")
True
>>> omc.sendExpression("simulate(OET.Example.multibodyWEC)")
{'resultFile': 'c:/OPENMODELICAGIT/OpenModelica/OMPython/OET.Example.multibodyWEC_res.mat', 'simulationOptions': "startTime = 0.0, stopTime = 500.0, numberOfIntervals = 10000, tolerance = 1e-8, method = 'dassl', fileNamePrefix = 'OET.Example.multibodyWEC', options = '', outputFormat = 'mat', variableFilter = '.*', cflags = '', simflags = ''", 'messages': 'LOG_STDOUT        | info    | ... loading "hydro.parameters.rho" from "C:/OPENMODELICAGIT/matlab/OceanEngineeringToolbox/RM3HydroCoeff.mat"\nLOG_STDOUT        | info    | ... loading "hydro.coefficients.excitation.spectralDecomp.im1" from "C:/OPENMODELICAGIT/matlab/OceanEngineeringToolbox/RM3HydroCoeff.mat"\nLOG_STDOUT        | info    | ... loading "hydro.coefficients.excitation.spectralDecomp.re1" from "C:/OPENMODELICAGIT/matlab/OceanEngineeringToolbox/RM3HydroCoeff.mat,  | info    | The simulation finished successfully.\n', 'timeFrontend': 2.8870213, 'timeBackend': 189.6112329, 'timeSimCode': 7.6194894, 'timeTemplates': 4.1510398, 'timeCompile': 26.2336279, 'timeSimulation': 9.5957187, 'timeTotal': 240.0990083}
```

So the first suggestion is please install the latest OpenModelica version from nightly builds and I hope that should automatically solve the problem.

**Reply by thomash303** [reply]
@arun3688  Okay I will give that a shot, that is optimistic!! 

Here is an example of the output it would return for me

![image](https://github.com/user-attachments/assets/2f925941-59ba-4395-aeba-e5c98baae988)

**Reply by arun3688** [reply]
@thomash303 I guess i can reproduce the issue when using `ModelicaSystem` constructor, I guess we are failing to add the missing dependencies to the constructor

**Reply by thomash303** [reply]
@arun3688 ah okay makes a lot more sense. How would you recommend I proceed? Use the scripting method that you performed originally?

**Reply by arun3688** [reply]
@thomash303 for the time being you use the script interface i sent until we find the solution, I will try to debug more and see whats happening

**Reply by thomash303** [reply]
@arun3688 Thank you for your help!
